lib/auth/auth_with_roles.go

```diff
@@ -1167,21 +1167,24 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
 	var roles []string
 	var traits wrappers.Traits
 
-	switch {
-	case a.hasBuiltinRole(string(teleport.RoleAdmin)):
-		// If it's an admin generating the certificate, the roles and traits for
-		// the user have to be fetched from the backend. This should be safe since
-		// this is typically done against a local user.
-		user, err := a.GetUser(req.Username, false)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		roles = user.GetRoles()
-		traits = user.GetTraits()
-	case req.Username == a.context.User.GetName():
-		// user is requesting TTL for themselves,
-		// limit the TTL to the duration of the session, to prevent
-		// users renewing their certificates forever
+	// this prevents clients who have no chance at getting a cert and impersonating anyone
+	// from enumerating local users and hitting database
+	if !a.hasBuiltinRole(string(teleport.RoleAdmin)) && !a.context.Checker.CanImpersonateSomeone() && req.Username != a.context.User.GetName() {
+		return nil, trace.AccessDenied("access denied")
+	}
+
+	// Extract the user and role set for whom the certificate will be generated.
+	// This should be safe since
+	// this is typically done against a local user.
+	user, err := a.GetUser(req.Username, false)
+	if err != nil {
+		return nil, trace.AccessDenied("access denied")
```

Comment:
Since you're replacing the original error with AccessDenied, I'd at least log `err` here so it's not lost.